torchvision/transforms/_core.py

```diff
@@ -0,0 +1,109 @@
+import collections.abc
+import inspect
+import re
+from typing import Any, Callable, Dict, Iterator, Optional, Type, TypeVar
+
+import torch
+from torch import nn
+
+from torchvision.features import BoundingBox, Feature, Image
+
+__all__ = ["Transform", "Compose", "query_sample"]
+
+T = TypeVar("T")
+
+
+def query_sample(sample: Any, fn: Callable[[Any], Optional[T]]) -> Iterator[T]:
+    if isinstance(sample, (collections.abc.Sequence, collections.abc.Mapping)):
+        for item in sample.values() if isinstance(sample, collections.abc.Mapping) else sample:
+            yield from query_sample(item, fn)
+    else:
+        result = fn(sample)
+        if result is not None:
+            yield result
+
+
+class Transform(nn.Module):
+    _KNOWN_FEATURE_TYPES = {
+        "_".join([part.lower() for part in re.findall("[A-Z][^A-Z]*", feature_type.__name__)]): feature_type
+        for feature_type in (Image, BoundingBox)
+    }
+
+    def __init_subclass__(cls, *, auto_register: bool = True):
+        cls._feature_transforms: Dict[Type[Feature], Callable] = {}
+        if auto_register:
+            cls._auto_register()
+
+    @classmethod
+    def _auto_register(cls) -> None:
+        for name, value in inspect.getmembers(cls):
+            # attribute is private
+            if name.startswith("_"):
+                continue
+
+            # attribute name does not match any known feature
+            try:
+                feature_type = cls._KNOWN_FEATURE_TYPES[name]
+            except KeyError:
+                continue
+
+            # attribute is not a static method
+            if not inspect.isfunction(value):
+                continue
+            argspec = inspect.getfullargspec(value)
+
+            # transform takes none or more than one than one positional arguments
+            if len(argspec.args) != 1:
+                continue
```

Comment:
As we discussed in VC, I believe this can lead to subtle bugs. If the user writes their transform as
```python
class Rotate(Transform):
    @staticmethod
    def image(img, angle):
        pass
```
instead of
```python
class Rotate(Transform):
    @staticmethod
    def image(img, *, angle):
        pass
```
then the transformation will not be automatically registered, and no errors / warnings will show up, only that the transform will silently not be applied because it wasn't properly auto-registered.

I think it might be better to be explicit here and instead use decorators to register a function to a transform. Something in the lines of
```
class Rotate(Transform):
    pass

@register_transform(transform=Rotate, input_type=Image)
def rotate_image(img, angle):
    pass
```

Reply:
> I think it might be better to be explicit here and instead use decorators to register a function to a transform.

1. I dislike this approach, because we would have to manually namespace everything with the feature type. For example, I think it is more concise to have `Rotate.image` and `Rotate.bounding_box` rather than `rotate_image` and `rotate_bounding_box`. Note that in the latter case, the namespace would have roughly n times more entries where n is the number of features we support.
2. The auto registering can be disabled and the registering can be performed manually in the constructor:

    ```python
    class Rotate(Transform, auto_register=False):
        def __init__(self):
            super().__init__()
            self.register_feature_transform(Image, self.foo)
    
        @staticmethod
        def foo(input, angle):
            pass
    ```
3. If we ultimately go with registering standalone functions with a decorator, I suggest we use a classmethod of the transform, i.e.

    ```python
    class Rotate(Transform):
        pass
    
    
    @Rotate.register(Image)
    def rotate_image(input, angle):
        pass
    ```

> As we discussed in VC, I believe this can lead to subtle bugs.

True, good catch! The rule can be made more robust though. We evoke the feature transforms with `feature_transform(input, **params)` where `params` can be an empty dict. That means all of these signatures need to be supported:

```python
def foo(input):
    pass

def foo(input=None):
    pass

def foo(input, bar):  # your example
    pass

def foo(input, bar=None):
    pass

def foo(input=None, bar=None):
    pass

def foo(input, *, bar):
    pass

def foo(input, *, bar=None):
    pass

def foo(input=None, *, bar):
    pass

def foo(input=None, *, bar=None):
    pass
```

For Python `>=3.8` we also need to keep positional-only arguments in mind. 

For the debugging, the user can use `transform.is_supported` and pass an feature type or instance to see if it would be transformed or not. But I agree that one would probably only do that after discovering that something is wrong. Another idea would be to add a `verbose` flag that prints extra information during the auto registration.

---

Since all of this is a trade-off between convenience with possible subtle bugs and explicitness with a higher verbosity, I think we should wait for more opinions before we decide.

Reply:
I've added a commit that makes the matching more robust and adds the option to print information about the process:

```python
class TestTransform(Transform, verbose=True):
    @staticmethod
    def incompatible_signature1():
        pass

    @staticmethod
    def incompatible_signature2(*, input):
        pass

    # Python >=3.8 only
    # @staticmethod
    # def incompatible_signature3(input, foo, /):
    #     pass

    @staticmethod
    def _private(input):
        pass

    @staticmethod
    def unknown(input):
        pass

    @staticmethod
    def imaeg(input):
        pass

    @staticmethod
    def image(input, foo):
        pass

    @staticmethod
    def boundingbox(input):
        pass

    @staticmethod
    def bounding_box(input, *, foo):
        pass
```

```
TestTransform._private() was not registered as feature transform, because it is private.
TestTransform.bounding_box() was registered as feature transform for type 'BoundingBox'.
TestTransform.boundingbox() was not registered as feature transform, because its name doesn't match any known feature type. Did you mean to name it 'bounding_box' to be registered for type 'BoundingBox'?
TestTransform.imaeg() was not registered as feature transform, because its name doesn't match any known feature type. Did you mean to name it 'image' to be registered for type 'Image'?
TestTransform.image() was registered as feature transform for type 'Image'.
TestTransform.incompatible_signature1() was not registered as feature transform, because it cannot be invoked with incompatible_signature1(input, **params).
TestTransform.incompatible_signature2() was not registered as feature transform, because it cannot be invoked with incompatible_signature2(input, **params).
TestTransform.unknown() was not registered as feature transform, because its name doesn't match any known feature type.
```